src/store.rs

```diff
@@ -132,668 +45,56 @@ impl Default for StringSet {
 
 impl StringSet {
     #[allow(dead_code)]
-    pub(crate) fn all_strings(&self) -> Vec<&str> {
+    pub fn all_strings(&self) -> Vec<&str> {
         self.0.into_iter().map(|x| x.1).collect::<Vec<_>>()
     }
 }
 
 #[derive(Debug)]
 pub struct Store<F: LurkField> {
-    pub(crate) cons_store: IndexSet<(Ptr<F>, Ptr<F>)>,
-    pub(crate) comm_store: IndexSet<(FWrap<F>, Ptr<F>)>,
+    pub cons_store: IndexSet<(Ptr<F>, Ptr<F>)>,
+    pub comm_store: IndexSet<(FWrap<F>, Ptr<F>)>,
 
-    fun_store: IndexSet<(Ptr<F>, Ptr<F>, Ptr<F>)>,
+    pub fun_store: IndexSet<(Ptr<F>, Ptr<F>, Ptr<F>)>,
 
-    pub(crate) sym_store: StringSet,
+    pub sym_store: StringSet,
 
     // Other sparse storage format without hashing is likely more efficient
-    pub(crate) num_store: IndexSet<Num<F>>,
-
-    str_store: StringSet,
-    thunk_store: IndexSet<Thunk<F>>,
-    call0_store: IndexSet<(Ptr<F>, ContPtr<F>)>,
-    call_store: IndexSet<(Ptr<F>, Ptr<F>, ContPtr<F>)>,
-    call2_store: IndexSet<(Ptr<F>, Ptr<F>, ContPtr<F>)>,
-    tail_store: IndexSet<(Ptr<F>, ContPtr<F>)>,
-    lookup_store: IndexSet<(Ptr<F>, ContPtr<F>)>,
-    unop_store: IndexSet<(Op1, ContPtr<F>)>,
-    binop_store: IndexSet<(Op2, Ptr<F>, Ptr<F>, ContPtr<F>)>,
-    binop2_store: IndexSet<(Op2, Ptr<F>, ContPtr<F>)>,
-    if_store: IndexSet<(Ptr<F>, ContPtr<F>)>,
-    let_store: IndexSet<(Ptr<F>, Ptr<F>, Ptr<F>, ContPtr<F>)>,
-    letrec_store: IndexSet<(Ptr<F>, Ptr<F>, Ptr<F>, ContPtr<F>)>,
-    emit_store: IndexSet<ContPtr<F>>,
-
-    opaque_map: dashmap::DashMap<Ptr<F>, ScalarPtr<F>>,
+    pub num_store: IndexSet<Num<F>>,
+
+    pub str_store: StringSet,
+    pub thunk_store: IndexSet<Thunk<F>>,
+    pub call0_store: IndexSet<(Ptr<F>, ContPtr<F>)>,
+    pub call_store: IndexSet<(Ptr<F>, Ptr<F>, ContPtr<F>)>,
+    pub call2_store: IndexSet<(Ptr<F>, Ptr<F>, ContPtr<F>)>,
+    pub tail_store: IndexSet<(Ptr<F>, ContPtr<F>)>,
+    pub lookup_store: IndexSet<(Ptr<F>, ContPtr<F>)>,
+    pub unop_store: IndexSet<(Op1, ContPtr<F>)>,
+    pub binop_store: IndexSet<(Op2, Ptr<F>, Ptr<F>, ContPtr<F>)>,
+    pub binop2_store: IndexSet<(Op2, Ptr<F>, ContPtr<F>)>,
+    pub if_store: IndexSet<(Ptr<F>, ContPtr<F>)>,
+    pub let_store: IndexSet<(Ptr<F>, Ptr<F>, Ptr<F>, ContPtr<F>)>,
+    pub letrec_store: IndexSet<(Ptr<F>, Ptr<F>, Ptr<F>, ContPtr<F>)>,
+    pub emit_store: IndexSet<ContPtr<F>>,
+
+    pub opaque_ptrs: IndexSet<ScalarPtr<F>>,
+    pub opaque_cont_ptrs: IndexSet<ScalarContPtr<F>>,
+
     /// Holds a mapping of ScalarPtr -> Ptr for reverse lookups
-    pub(crate) scalar_ptr_map: dashmap::DashMap<ScalarPtr<F>, Ptr<F>, ahash::RandomState>,
+    pub scalar_ptr_map: dashmap::DashMap<ScalarPtr<F>, Ptr<F>, ahash::RandomState>,
```

Comment:
Could `scalar_ptr_map` and `scalar_cont_ptr_map` be merged given they have the same `SPtr` representation? (same for `opaque_ptrs`/`opaque_cont_ptrs` above)

Reply:
I'm actually a lot more circumspect about this. The DashMap seem to be somewhat lock-contended, see:
```
src/store.rs
1925-
1926-    // Get hash for expr, but only if it already exists. This should never cause create_scalar_ptr to be called. Use
1927:    // this after the cache has been hydrated. NOTE: because dashmap::entry can deadlock, it is important not to call
1928-    // hash_expr in nested call graphs which might trigger that behavior. This discovery is what led to get_expr_hash
1929-    // and the 'get' versions of hash_cons, hash_sym, etc.
1930-    pub fn get_expr_hash(&self, ptr: &Ptr<F>) -> Option<ScalarPtr<F>> {
1931-        self.hash_expr_aux(ptr, HashScalar::Get)
```

I would trust @porcuquine to write an issue highlighting the risks though?

Reply:
Done, regarding the change here. I think merging the maps should be fine, since collisions should be impossible.